File: tools/icloud_photo_sync/photos_bridge.swift
import Foundation
import Photos

struct ResourcePayload: Encodable {
    let asset_local_identifier: String
    let asset_uuid: String
    let media_type: Int
    let media_subtypes: UInt64
    let creation_date: String?
    let modification_date: String?
    let resource_index: Int
    let resource_type: Int
    let original_filename: String
    let uniform_type_identifier: String?
    let file_size: Int64?
}

struct ExportPayload: Encodable {
    let output_path: String
}

struct BatchRequest: Decodable {
    let request_id: String
    let asset_local_identifier: String
    let resource_index: Int
    let output_path: String
}

struct BatchResponse: Encodable {
    let request_id: String
    let output_path: String?
    let error: String?
}

func requestAuthorization() {
    let semaphore = DispatchSemaphore(value: 0)
    var status: PHAuthorizationStatus = .notDetermined
    PHPhotoLibrary.requestAuthorization(for: .readWrite) { newStatus in
        status = newStatus
        semaphore.signal()
    }
    _ = semaphore.wait(timeout: .now() + 60)
    guard status == .authorized || status == .limited else {
        fputs("photos authorization unavailable: \(status.rawValue)\n", stderr)
        exit(1)
    }
}

func isoString(_ date: Date?) -> String? {
    guard let date else { return nil }
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter.string(from: date)
}

func fileSize(of resource: PHAssetResource) -> Int64? {
    if let number = resource.value(forKey: "fileSize") as? NSNumber {
        return number.int64Value
    }
    if let value = resource.value(forKey: "fileSize") as? NSString {
        return Int64(value as String)
    }
    return nil
}

func encodeLine<T: Encodable>(_ payload: T) {
    let encoder = JSONEncoder()
    guard let data = try? encoder.encode(payload),
          let line = String(data: data, encoding: .utf8) else {
        fputs("json encode failed\n", stderr)
        exit(1)
    }
    print(line)
}

func fetchAsset(localIdentifier: String) -> PHAsset {
    let result = PHAsset.fetchAssets(withLocalIdentifiers: [localIdentifier], options: nil)
    guard let asset = result.firstObject else {
        fputs("asset not found: \(localIdentifier)\n", stderr)
        exit(1)
    }
    return asset
}

func listResources() {
    let assets = PHAsset.fetchAssets(with: nil)
    assets.enumerateObjects { asset, _, _ in
        let assetUUID = asset.localIdentifier.split(separator: "/").first.map(String.init) ?? asset.localIdentifier
        let resources = PHAssetResource.assetResources(for: asset)
        for (index, resource) in resources.enumerated() {
            let payload = ResourcePayload(
                asset_local_identifier: asset.localIdentifier,
                asset_uuid: assetUUID,
                media_type: asset.mediaType.rawValue,
                media_subtypes: UInt64(asset.mediaSubtypes.rawValue),
                creation_date: isoString(asset.creationDate),
                modification_date: isoString(asset.modificationDate),
                resource_index: index,
                resource_type: resource.type.rawValue,
                original_filename: resource.originalFilename,
                uniform_type_identifier: resource.uniformTypeIdentifier,
                file_size: fileSize(of: resource)
            )
            encodeLine(payload)
        }
    }
}

func exportResource(localIdentifier: String, resourceIndex: Int, outputPath: String) {
    let asset = fetchAsset(localIdentifier: localIdentifier)
    let resources = PHAssetResource.assetResources(for: asset)
    guard resourceIndex >= 0 && resourceIndex < resources.count else {
        fputs("resource index out of range\n", stderr)
        exit(1)
    }
    let resource = resources[resourceIndex]
    let outputURL = URL(fileURLWithPath: outputPath)
    try? FileManager.default.removeItem(at: outputURL)
    try? FileManager.default.createDirectory(at: outputURL.deletingLastPathComponent(), withIntermediateDirectories: true)
    let options = PHAssetResourceRequestOptions()
    options.isNetworkAccessAllowed = true

    let semaphore = DispatchSemaphore(value: 0)
    var exportError: Error?
    PHAssetResourceManager.default().writeData(for: resource, toFile: outputURL, options: options) { error in
        exportError = error
        semaphore.signal()
    }
    _ = semaphore.wait(timeout: .now() + 3600)
    if let exportError {
        fputs("\(exportError)\n", stderr)
        exit(1)
    }
    encodeLine(ExportPayload(output_path: outputURL.path))
}

func exportBatch(requestFile: String) {
    let requestURL = URL(fileURLWithPath: requestFile)
    let data: Data
    do {
        data = try Data(contentsOf: requestURL)
    } catch {
        fputs("failed to read batch request file\n", stderr)
        exit(1)
    }

    let requests: [BatchRequest]
    do {
        requests = try JSONDecoder().decode([BatchRequest].self, from: data)
    } catch {
        fputs("failed to decode batch request json\n", stderr)
        exit(1)
    }

    for request in requests {
        let assetResult = PHAsset.fetchAssets(withLocalIdentifiers: [request.asset_local_identifier], options: nil)
        guard let asset = assetResult.firstObject else {
            encodeLine(BatchResponse(request_id: request.request_id, output_path: nil, error: "asset not found"))
            continue
        }
        let resources = PHAssetResource.assetResources(for: asset)
        guard request.resource_index >= 0 && request.resource_index < resources.count else {
            encodeLine(BatchResponse(request_id: request.request_id, output_path: nil, error: "resource index out of range"))
            continue
        }

        let resource = resources[request.resource_index]
        let outputURL = URL(fileURLWithPath: request.output_path)
        try? FileManager.default.removeItem(at: outputURL)
        try? FileManager.default.createDirectory(at: outputURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        let options = PHAssetResourceRequestOptions()
        options.isNetworkAccessAllowed = true

        let semaphore = DispatchSemaphore(value: 0)
        var exportError: Error?
        PHAssetResourceManager.default().writeData(for: resource, toFile: outputURL, options: options) { error in
            exportError = error
            semaphore.signal()
        }
        _ = semaphore.wait(timeout: .now() + 3600)
        if let exportError {
            encodeLine(BatchResponse(request_id: request.request_id, output_path: nil, error: "\(exportError)"))
            continue
        }
        encodeLine(BatchResponse(request_id: request.request_id, output_path: outputURL.path, error: nil))
    }
}

requestAuthorization()

let args = CommandLine.arguments
guard args.count >= 2 else {
    fputs("usage: photos_bridge <list-resources|export-resource> [...]\n", stderr)
    exit(2)
}

switch args[1] {
case "list-resources":
    listResources()
case "export-resource":
    guard args.count == 5, let resourceIndex = Int(args[3]) else {
        fputs("usage: photos_bridge export-resource <asset-local-id> <resource-index> <output-path>\n", stderr)
        exit(2)
    }
    exportResource(localIdentifier: args[2], resourceIndex: resourceIndex, outputPath: args[4])
case "export-batch":
    guard args.count == 3 else {
        fputs("usage: photos_bridge export-batch <request-json>\n", stderr)
        exit(2)
    }
    exportBatch(requestFile: args[2])
default:
    fputs("unknown command: \(args[1])\n", stderr)
    exit(2)
}

File: tools/icloud_photo_sync/photos_bridge.py
from __future__ import annotations

import json
import os
import subprocess
from dataclasses import dataclass
from pathlib import Path
from tempfile import NamedTemporaryFile


@dataclass(frozen=True)
class PhotosResourceDescriptor:
    asset_local_identifier: str
    asset_uuid: str
    media_type: int
    media_subtypes: int
    creation_date: str | None
    modification_date: str | None
    resource_index: int
    resource_type: int
    original_filename: str
    uniform_type_identifier: str | None
    file_size: int | None

    @property
    def resource_key(self) -> str:
        return f"{self.asset_uuid}:{self.resource_index}:{self.original_filename}"


class PhotosBridge:
    def __init__(self, swift_source: Path, build_dir: Path) -> None:
        self.swift_source = Path(swift_source)
        self.build_dir = Path(build_dir)
        self.binary_path = self.build_dir / "photos_bridge"

    def ensure_binary(self) -> Path:
        bundled_path = os.environ.get("PHOTO_STEWARD_PHOTOS_BRIDGE")
        if bundled_path:
            candidate = Path(bundled_path).expanduser()
            if candidate.is_file() and candidate.stat().st_mode & 0o111:
                return candidate

        self.build_dir.mkdir(parents=True, exist_ok=True)
        if self.binary_path.exists() and self.binary_path.stat().st_mtime >= self.swift_source.stat().st_mtime:
            return self.binary_path

        if not self.swift_source.exists():
            raise RuntimeError(
                "Photos bridge is not available in this installation; "
                "install the Photo Steward App or use a source checkout with Xcode"
            )

        command = [
            "swiftc",
            str(self.swift_source),
            "-o",
            str(self.binary_path),
            "-framework",
            "Photos",
        ]
        try:
            proc = subprocess.run(command, capture_output=True, text=True)
        except FileNotFoundError as exc:
            raise RuntimeError(
                "Photos bridge is not bundled and swiftc is unavailable; "
                "install the Photo Steward App or use a source checkout with Xcode"
            ) from exc
        if proc.returncode != 0:
            raise RuntimeError(proc.stderr.strip() or proc.stdout.strip() or "swiftc failed")
        return self.binary_path

    def _run(self, *args: str) -> subprocess.CompletedProcess[str]:
        binary = self.ensure_binary()
        proc = subprocess.run(
            [str(binary), *args],
            capture_output=True,
            text=True,
        )
        if proc.returncode != 0:
            raise RuntimeError(proc.stderr.strip() or proc.stdout.strip() or "photos bridge failed")
        return proc

    def list_resources(self) -> list[PhotosResourceDescriptor]:
        proc = self._run("list-resources")
        resources: list[PhotosResourceDescriptor] = []
        for line in proc.stdout.splitlines():
            if not line.strip():
                continue
            payload = json.loads(line)
            resources.append(
                PhotosResourceDescriptor(
                    asset_local_identifier=payload["asset_local_identifier"],
                    asset_uuid=payload["asset_uuid"],
                    media_type=int(payload["media_type"]),
                    media_subtypes=int(payload["media_subtypes"]),
                    creation_date=payload.get("creation_date"),
                    modification_date=payload.get("modification_date"),
                    resource_index=int(payload["resource_index"]),
                    resource_type=int(payload["resource_type"]),
                    original_filename=payload["original_filename"],
                    uniform_type_identifier=payload.get("uniform_type_identifier"),
                    file_size=payload.get("file_size"),
                )
            )
        return resources

    def export_resource(self, asset_local_identifier: str, resource_index: int, output_path: Path) -> Path:
        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        proc = self._run("export-resource", asset_local_identifier, str(resource_index), str(output_path))
        payload = json.loads(proc.stdout.strip() or "{}")
        exported = Path(payload.get("output_path", output_path))
        if not exported.exists():
            raise RuntimeError(f"exported file missing: {exported}")
        return exported

    def export_resources_batch(self, requests: list[dict]) -> dict[str, dict]:
        if not requests:
            return {}

        with NamedTemporaryFile("w", encoding="utf-8", suffix=".json", delete=False) as handle:
            json.dump(requests, handle, ensure_ascii=False)
            request_path = Path(handle.name)

        try:
            proc = self._run("export-batch", str(request_path))
        finally:
            request_path.unlink(missing_ok=True)

        results: dict[str, dict] = {}
        for line in proc.stdout.splitlines():
            if not line.strip():
                continue
            payload = json.loads(line)
            results[payload["request_id"]] = payload
        return results
